File: Binary Search/5. Maxval/Maxval.py
import sys, bisect, random

sys.stdin = open("Maxval.inp", "r")
sys.stdout = open("Maxval.out", "w")

n = int(input())
a = list(map(int, input().split())) + [0]
a[:-1].sort()

print(1000000)
for i in range(2*pow(10, 6) + 1):
    print(random.randint(1, 1000000), end = ' ')

"""
r = 0
for i in range(n):
    lim = bisect.bisect(a, 2*a[i])
    lim = min(n+1, lim+1)
    
    for j in range(i+1, lim+1):
        if j >= n:
            break
        r = max(r, a[j] % a[i])
print(r)"""

File: Binary Search/5. Maxval/Maxval.cpp
#include <iostream>
#include <fstream>
#include <algorithm>

long long n, a[200001], r;

int main () {
    std::ios_base::sync_with_stdio(NULL); std::cin.tie(NULL); std::cout.tie(NULL);
    freopen("Maxval.inp", "r", stdin);
    freopen("Maxval.out", "w", stdout);
    
    std::cin >> n;
    for (int i = 0; i < n; i++) std::cin >> a[i];
    std::sort(a, a+n);

    for (int i = 0; i < n; i++) {
        r = std::max(r, a[std::lower_bound(a, a+n, 2*a[i]) - a - 1] % a[i]);
    }
    std::cout << r << '\n';
}